wvd-templates/DSC/Functions.ps1

```diff
@@ -327,4 +330,163 @@ function IsRDAgentRegistryValidForRegistration {
+
+.PARAMETER AgentInstallerFolder
+Required path to MSI installer file
+
+.PARAMETER AgentBootServiceInstallerFolder
+Required path to MSI installer file
+
+.PARAMETER StartAgent
+Start the agent service (RDAgentBootLoader) immediately
+#>
+function InstallRDAgents {
+    Param(
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [string]$AgentInstallerFolder,
+        
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [string]$AgentBootServiceInstallerFolder,
+    
+        [Parameter(Mandatory = $false)]
+        [ValidateNotNullOrEmpty()]
+        [string]$RegistrationToken,
+    
+        [Parameter(Mandatory = $false)]
+        [ValidateNotNullOrEmpty()]
+        [bool]$StartAgent,
+    
+        [Parameter(mandatory = $false)]
+        [switch]$EnableVerboseMsiLogging
+    )
+
+    # Convert relative paths to absolute paths if needed
+    Write-Log -Message "Boot loader folder is $AgentBootServiceInstallerFolder"
+    $AgentBootServiceInstaller = (Get-ChildItem $AgentBootServiceInstallerFolder\ -Filter *.msi | Select-Object).FullName
+    if ((-not $AgentBootServiceInstaller) -or (-not (Test-Path $AgentBootServiceInstaller))) {
+        throw "RD Infra Agent Installer package is not found '$AgentBootServiceInstaller'"
+    }
+
+    # Convert relative paths to absolute paths if needed
+    Write-Log -Message "Agent folder is $AgentInstallerFolder"
+    $AgentInstaller = (Get-ChildItem $AgentInstallerFolder\ -Filter *.msi | Select-Object).FullName
+    if ((-not $AgentInstaller) -or (-not (Test-Path $AgentInstaller))) {
+        throw "RD Infra Agent Installer package is not found '$AgentInstaller'"
+    }
+
+    if (!$RegistrationToken) {
+        throw "No registration token specified"
+    }
+
+    RunMsiWithRetry -programDisplayName "RDAgentBootLoader" -isUninstall -argumentList @("/x {A38EE409-424D-4A0D-B5B6-5D66F20F62A5}", "/quiet", "/qn", "/norestart", "/passive") -msiOutputLogPath "C:\Users\AgentBootLoaderUnInstall.txt" -msiLogVerboseOutput:$EnableVerboseMsiLogging
+
+    while ($true) {
+        try {
+            $oldAgent = Get-Package -ProviderName msi -Name "Remote Desktop Services Infrastructure Agent" -ErrorAction Stop
+        }
+        catch {
+            #Ignore the error if it was due to no packages being found.
+            if ($_.FullyQualifiedErrorId -eq "NoMatchFound,Microsoft.PowerShell.PackageManagement.Cmdlets.GetPackage") {
+                break
+            }
+
+            throw;
+        }
+
+        $oldVersion = $oldAgent.Version
+        $productCodeParameter = $oldAgent.FastPackageReference
+
+        RunMsiWithRetry -programDisplayName "RD Infra Agent $oldVersion" -isUninstall -argumentList @("/x $productCodeParameter", "/quiet", "/qn", "/norestart", "/passive") -msiOutputLogPath "C:\Users\AgentUninstall.txt" -msiLogVerboseOutput:$EnableVerboseMsiLogging

```

Comment:
Wouldn't this be an infinite loop if this keeps on timing out because of ERROR_INSTALL_ALREADY_RUNNING ? Eventually the DSC ext will timeout

Reply:
Good catch I will update RunMsiWithRetry to also throw an exception for both the install and uninstall case when the ERROR_INSTALL_ALREADY_RUNNING  retry limit is reached.